README.md
# Rest Navigator

A simple flutter router that parses path parameters.

This package was inspired by [fluro](https://github.com/theyakka/fluro).

## Getting Started

The easiest way to get started is by looking at an example:

```dart
import 'package:flutter/material.dart';
import 'package:rest_router/parser.dart';
import 'package:rest_router/rest_navigator.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  final router = Router({
    "/": (BuildContext context, parameters) => HomePage(),
    "/items": (BuildContext context, parameters) => SubPage("path is /items"),
    "/plus_two/:num": (BuildContext context, parameters) => SubPage("sum is ${int.parse(parameters["num"]?.first) + 2}"),
  }, onUnknownRouteHandler: (BuildContext context, parameters) => SubPage("${parameters[urlPathKey]?.first} is 404"));

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Rest navigator demo',
      initialRoute: "/",
      onGenerateRoute: router.generator,
    );
  }
}

/// Navigates to sub pages.
class HomePage extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: Text("/")),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            RaisedButton(
              child: Text("Item list"),
              onPressed: () => Navigator.of(context).pushNamed("/items", arguments: TransitionType.native),
            ),
            RaisedButton(
              child: Text("Run /plus_two/40"),
              onPressed: () => Navigator.of(context).pushNamed("/plus_two/40", arguments: TransitionType.native),
            ),
            RaisedButton(
              child: Text("Invalid path"),
              onPressed: () => Navigator.of(context).pushNamed("/itemz"), // Defaults to instant transition
            ),
          ],
        ),
      ),
    );
  }
}

/// Scaffold with app bar title.
class SubPage extends StatelessWidget {
  final String title;

  SubPage(this.title);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: Text(title)),
    );
  }
}
```

Parameters is a `Map<String, List<String>>` to allow multiple variables of the same name.

Note you have specify `arguments: TransitionType.native` to achieve native transitions. The rationale is that this is better for a deep-link powered app. Feel free to fork this project to make opiniated changes.

The [source code](https://github.com/Dennis-Krasnov/Flutter-Rest-Router) is available to fully understand what's going in under the hood, and PRs are always welcome.

CHANGELOG.md
## [0.1.0]
* Removed debugging statements.

## [0.0.1]

* Created basic router for flutter.
* Added native and no route transitions.
* Added simple path parameter parsing.
* Tested widgets and parser.


test/parser_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:rest_router/parser.dart';

void main() {
  group("Successfuly parse valid URLs", () {
    test("Parse normal URL", () {
      expect(
        parsePathParameters("/settings/theme", "/settings/theme"),
        <String, List<String>>{
          handlerPathKey: ["/settings/theme"],
          urlPathKey: ["/settings/theme"],
        }
      );
    });

    test("Parse URL with path variables", () {
      expect(
        parsePathParameters("/user/:id", "/user/323"),
        <String, List<String>>{
          handlerPathKey: ["/user/:id"],
          urlPathKey: ["/user/323"],
          "id": ["323"],
        }
      );
    });

    test("Parse URL with repeating path variables", () {
      expect(
        parsePathParameters("user/:id/again/:id", "user/123/again/456"),
        <String, List<String>>{
          handlerPathKey: ["user/:id/again/:id"],
          urlPathKey: ["user/123/again/456"],
          "id": ["123", "456"],
        }
      );
    });

    // TODO: extract variables such as ?var=value&another=value2 as well!
  });

  group("Fail to parse malformed URLs", () {
    test("Mismatched parameter count", () {
      expect(
        () => parsePathParameters("/settings/theme", "settings"),
        throwsArgumentError
      );
    });

    test("Mismatched parameters", () {
      expect(
          () => parsePathParameters("/settings/theme", "/settings/themes"),
          throwsArgumentError
      );
    });
  });
}

// Map<String, List<String>> parsePathParameters(String handlerPath, String url) {

example/test/navigator_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:example/main.dart';

void main() {
  testWidgets("Navigates to static path", (WidgetTester tester) async {
    await tester.pumpWidget(MyApp());
    expect(find.text("/"), findsOneWidget);

    await tester.tap(find.byType(RaisedButton).first);
    await tester.pumpAndSettle();
    expect(find.text("path is /items"), findsOneWidget);

    await tester.tap(find.bySemanticsLabel("Back"));
    await tester.pumpAndSettle();
    expect(find.text("/"), findsOneWidget);
  });

  testWidgets("Navigates to variable path", (WidgetTester tester) async {
    await tester.pumpWidget(MyApp());
    expect(find.text("/"), findsOneWidget);

    await tester.tap(find.byType(RaisedButton).at(1));
    await tester.pumpAndSettle();
    expect(find.text("sum is 42"), findsOneWidget);

    await tester.tap(find.bySemanticsLabel("Back"));
    await tester.pumpAndSettle();
    expect(find.text("/"), findsOneWidget);
  });

  testWidgets("Navigates to 404", (WidgetTester tester) async {
    await tester.pumpWidget(MyApp());
    expect(find.text("/"), findsOneWidget);

    await tester.tap(find.byType(RaisedButton).at(2));
    await tester.pumpAndSettle();
    expect(find.text("/itemz is 404"), findsOneWidget);

    await tester.tap(find.bySemanticsLabel("Back"));
    await tester.pumpAndSettle();
    expect(find.text("/"), findsOneWidget);
  });
}


lib/parser.dart
// TODO: rename something more descriptive, place under router?
import 'package:rest_router/rest_navigator.dart';

final handlerPathKey = "__HANDLER_PATH__";
final urlPathKey = "__PATH__";

/// Extracts path variables from a [url] given the [handlerPath].
/// After creating this, I realized https://github.com/peterzagoranski/flutter_router implemented this better.
Map<String, List<String>> parsePathParameters(String handlerPath, String url) {
  // Don't parse unknown routes
  if (handlerPath == Router.unknownRoute) {
    return <String, List<String>>{
      handlerPathKey: [handlerPath],
      urlPathKey: [url]
    };
  }

  // Sections of handler URL separated by slashes
  final handlerParts = handlerPath.split("/");

  // Sections of URL separated by slashes
  final parts = url.split("/");

  if (parts.length != handlerParts.length) {
    throw ArgumentError("Part count didn't match");
  }

  // Path variables
  var parameters = <String, List<String>>{
    handlerPathKey: [handlerPath],
    urlPathKey: [url]
  };

  // Lockstep
  for(int i = 0; i < handlerParts.length; i++) {
    // Variables only
    if (handlerParts[i].startsWith(":")) {
      // Variable name
      final key = handlerParts[i].substring(1);

      // Defensively create list for key
      if (!parameters.containsKey(key)) {
        parameters[key] = [];
      }

      // Add variable value to list
      parameters[key].add(parts[i]);
    }
    else if (parts[i] != handlerParts[i]) {
      throw ArgumentError("Parts didn't match");
    }
  }

  return parameters;
}

lib/route_generator.dart
import 'package:flutter/material.dart';

import 'package:rest_router/parser.dart';
import 'package:rest_router/rest_navigator.dart';

/// Builder function with [parameters].
typedef Widget HandlerFunc(BuildContext context, Map<String, List<String>> parameters);

class Router {
  static final unknownRoute = "__UKNOWN_ROUTE__";

  /// Path with variables -> widget builders
  /// eg. {"days/:day": (BuildContext context, Map<String, List<String>> parameters) => Container()}
  final Map<String, HandlerFunc> pathHandlers;

  /// Builder for unknown route
  final HandlerFunc onUnknownRouteHandler;

  Router(this.pathHandlers, {@required this.onUnknownRouteHandler}) {
    pathHandlers[unknownRoute] = onUnknownRouteHandler;
  }

  /// Helper to find a single path handler with matching path parameters.
  bool _matchPathHandler(List<String> parts, String handlerPath) {
    // Sections of handler URL separated by slashes
    final handlerParts = handlerPath.split("/");

    // Must have same number of parts
    if (parts.length != handlerParts.length) {
      return false;
    }

    // Lockstep
    for (int i = 0; i < parts.length; i++) {
      // Non-variable parts must match
      if (!handlerParts[i].startsWith(":") && parts[i] != handlerParts[i]) {
        return false;
      }
    }

    // Route found
    return true;
  }

  /// Generates routes based on [routeSettings].
  /// Used with [MaterialApp.onGenerateRoute] and normal [Navigator] operations.
  Route<dynamic> generator(RouteSettings routeSettings) {
    // Determines route transition based on navigator arguments
    // Defaults to no transition, may throw cast error
    final transitionType = routeSettings.arguments as TransitionType ?? TransitionType.none;

    // Sections of URL separated by slashes
    final parts = routeSettings.name.split("/");

    // Find corresponding path handler
    // Fails if multiple handlers found
    // Defaults to [onUnknownRouteHandler] if no handlers found
    final handlerPath = pathHandlers.keys.singleWhere((handlerPath) => _matchPathHandler(parts, handlerPath), orElse: () => unknownRoute);

    // Parsed parameters and their values
    final parameters = parsePathParameters(handlerPath, routeSettings.name);

    // Navigate to the route
    switch(transitionType) {
      // Animation
      case TransitionType.native:
        return MaterialPageRoute<dynamic>(
          settings: routeSettings,
          builder: (BuildContext context) => pathHandlers[handlerPath](context, parameters),
        );
        break;
      // Instant
      case TransitionType.none:
        return NoAnimationMaterialPageRoute(
          settings: routeSettings,
          builder: (BuildContext context) => pathHandlers[handlerPath](context, parameters),
        );
        break;
      default:
        throw UnimplementedError("Invalid transition type: $transitionType");
    }
  }
}

example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:rest_router/parser.dart';
import 'package:rest_router/rest_navigator.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  final router = Router({
    "/": (BuildContext context, parameters) => HomePage(),
    "/items": (BuildContext context, parameters) => SubPage("path is /items"),
    "/plus_two/:num": (BuildContext context, parameters) => SubPage("sum is ${int.parse(parameters["num"]?.first) + 2}"),
  }, onUnknownRouteHandler: (BuildContext context, parameters) => SubPage("${parameters[urlPathKey]?.first} is 404"));

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Rest navigator demo',
      initialRoute: "/",
      onGenerateRoute: router.generator,
    );
  }
}

/// Navigates to sub pages.
class HomePage extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: Text("/")),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            RaisedButton(
              child: Text("Item list"),
              onPressed: () => Navigator.of(context).pushNamed("/items", arguments: TransitionType.native),
            ),
            RaisedButton(
              child: Text("Run /plus_two/40"),
              onPressed: () => Navigator.of(context).pushNamed("/plus_two/40", arguments: TransitionType.native),
            ),
            RaisedButton(
              child: Text("Invalid path"),
              onPressed: () => Navigator.of(context).pushNamed("/itemz"), // Defaults to instant transition
            ),
          ],
        ),
      ),
    );
  }
}

/// Scaffold with app bar title.
class SubPage extends StatelessWidget {
  final String title;

  SubPage(this.title);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: Text(title)),
    );
  }
}
